File: Manufacturing/src/Python/package/azuresphere_device_api/network.py
# Copyright (c) Microsoft Corporation. All rights reserved.
# Licensed under the MIT License.

from typing import List, Literal

import azuresphere_device_api.utils as utils
from azuresphere_device_api.exceptions import ValidationError
from azuresphere_device_api.validation import since_device_api_version

__all__ = ['configure_proxy', 'delete_network_proxy', 'get_all_failed_network_connections', 'get_all_network_interfaces', 'get_failed_network_connection',
           'get_network_firewall_ruleset', 'get_network_interface', 'get_network_proxy', 'get_network_status', 'set_network_interfaces']


def configure_proxy(
    enabled: bool,
    address: str,
    port: int,
    no_proxy_address: List[str],
    authentication_type: Literal["anonymous", "basic"],
    username=None,
    password=None,
) -> dict:
    """Makes a "POST" request to configure the network proxy on the attached device.

    :param enabled: Enable/disable network proxy on attached device.
    :type enabled: bool
    :param address: The network address of the proxy.
    :type address: str
    :param port: The port on the network proxy to be used.
    :type port: int
    :param no_proxy_address: Array of space-separated network addresses the device should avoid for
        proxy connection.
    :type no_proxy_address: list[str]
    :param authentication_type: If the proxy requires a user name and password, set this to basic,
        otherwise set it to anonymous.
    :type authentication_type: str
    :accepted-values authentication_type: anonymous, basic
    :param username: Optional parameter. Username used for proxy authentication in basic mode.
    :type username: str
    :param password: Optional parameter. Password used for proxy authentication in basic mode.
    :type password: str
    :return: The details of the supplied proxy configuration on success. An exception will be thrown on error.
    :rtype: dict[str, str]
    :raises: requests.exceptions
    :raises: AzureSphereDeviceApiException
    """

    if not address:
        raise ValidationError(
            "ERROR: Cannot configure proxy, address was null or empty.")

    valid_authentication_types = ["anonymous", "basic"]

    if authentication_type not in valid_authentication_types:
        raise ValidationError(
            "ERROR: Cannot configure proxy, invalid authentication type.")

    if username and password:

        if authentication_type == "anonymous":
            raise ValidationError(
                "ERROR: Cannot configure proxy, incorrect configuration of authentication type and proxy credentials.")

        # Basic type network proxy
        body = {
            "enabled": enabled,
            "address": address,
            "port": port,
            "noProxyAddresses": no_proxy_address,
            "authenticationType": authentication_type,
            "username": username,
            "password": password,
        }
        return utils.post_request("net/proxy", body)

    if authentication_type == "basic" or username or password:
        raise ValidationError(
            "ERROR: Cannot configure proxy, incorrect configuration of authentication type and proxy credentials.")

    # Anonymous type network proxy
    body = {
        "enabled": enabled,
        "address": address,
        "port": port,
        "noProxyAddresses": no_proxy_address,
        "authenticationType": authentication_type,
    }
    return utils.post_request("net/proxy", body)


def delete_network_proxy() -> dict:
    """Makes a "DELETE" request to delete the current network proxy.

    :return: An empty response on success. An exception will be thrown on error.
    :rtype: dict[str, str]
    :raises: requests.exceptions
    :raises: AzureSphereDeviceApiException
    """
    return utils.delete_request("net/proxy")


def get_all_failed_network_connections() -> dict:
    """Makes a "GET" request to get all the failed network connection attempts.

    :return: The failed connection attempts on success. An exception will be thrown on error.
    :rtype: dict[str, str]
    :raises: requests.exceptions
    :raises: AzureSphereDeviceApiException
    """
    return utils.get_request("wifi/diagnostics/networks")


@since_device_api_version("3.1.0")
def get_all_network_interfaces() -> dict:
    """Makes a "GET" request to get the status of all network interfaces.

    :return: The status of all network interfaces on success. An exception will be thrown on error.
    :rtype: dict[str, str]
    :raises: requests.exceptions
    :raises: AzureSphereDeviceApiException
    """
    return utils.get_request("net/interfaces")


def get_failed_network_connection(network_id: str) -> dict:
    """Makes a "GET" request to get a specific failed network connection attempt.

    :param network_id: The id of the network you would like the failed connection attempt of.
    :return: The failed network connection attempt on success. An exception will be thrown on error.
    :rtype: dict[str, str]
    :raises: requests.exceptions
    :raises: AzureSphereDeviceApiException
    """
    return utils.get_request(f"wifi/diagnostics/networks/{network_id}")


@since_device_api_version("4.1.0")
def get_network_firewall_ruleset() -> dict:
    """Makes a "GET" request to get all network firewall rulesets.

    :return: All network firewall rulesets on success. An exception will be thrown on error.
    :rtype: dict[str, str]
    :raises: requests.exceptions
    :raises: AzureSphereDeviceApiException
    """
    return utils.get_request("net/firewall/rulesets")


@since_device_api_version("3.1.0")
def get_network_interface(network_interface_name: str) -> dict:
    """Makes a "GET" request to get the status of the named network interface.

    :param network_interface_name: The name of the network interface you would like to get the
    status of.
    :type network_interface_name: str
    :return: The status of the named network interface on success. An exception will be thrown on error.
    :rtype: dict[str, str]
    :raises: requests.exceptions
    :raises: AzureSphereDeviceApiException
    """
    if not network_interface_name:
        raise ValidationError(
            "ERROR: Cannot get network interface, input was null or empty.")

    return utils.get_request(f"net/interfaces/{network_interface_name}")


def get_network_proxy() -> dict:
    """Makes a "GET" request to get information about the current network proxy.

    :return: The details of the current network proxy on success. An exception will be thrown on error.
    :rtype: dict[str, str]
    :raises: requests.exceptions
    :raises: AzureSphereDeviceApiException
    """
    return utils.get_request("net/proxy")


@since_device_api_version("3.1.0")
def get_network_status() -> dict:
    """Makes a "GET" request to get the network status.

    :return: The status of the current network on success. An exception will be thrown on error.
    :rtype: dict[str, str]
    :raises: requests.exceptions
    :raises: AzureSphereDeviceApiException
    """
    return utils.get_request("net/status")


@since_device_api_version("3.1.0")
def set_network_interfaces(
    network_interface_name: str, interface_up: bool
) -> dict:
    """Makes a "PATCH" request to add/update a specific network interface's attributes.

    :param network_interface_name: The name of the network interface you would like to add/update.
    :type network_interface_name: string
    :param interface_up: If you would like the interface to be active or not.
    :type interface_up: boolean
    :return: An empty response on success. An exception will be thrown on error.
    :rtype: dict[str, str]
    :raises: requests.exceptions
    :raises: AzureSphereDeviceApiException
    """

    if not network_interface_name:
        raise ValidationError(
            "ERROR: Cannot set network interface, network interface name was null or empty.")

    return utils.patch_request(
        f"net/interfaces/{network_interface_name}",
        {"interfaceUp": interface_up},
    )

File: Manufacturing/src/CSharp/Nuget/Package/Network.cs
/* Copyright (c) Microsoft Corporation. All rights reserved.
   Licensed under the MIT License. */
namespace Microsoft.Azure.Sphere.DeviceAPI
{
    using System.Collections.Generic;
    using System;

    /// <summary>
    /// Device REST APIs to provide information about the status of network interfaces on the attached device.
    /// </summary>
    public static class Network
    {
        /// <summary>Makes a "POST" request to configure the network proxy on the attached device.</summary>
        /// <param name="enabled">Enable network proxy on the attached device.</param>
        /// <param name="address">The network address of the proxy.</param>
        /// <param name="port">The port on the network address to be used.</param>
        /// <param name="noProxyAddresses">Array of space-separated network addresses the device should avoid for proxy connection.</param>
        /// <param name="authenticationType">If the proxy requires a user name and password, set this to basic, otherwise anonymous.</param>
        /// <param name="username">This is an optional parameter for the basic authentication version. The username used for proxy authentication.</param>
        /// <param name="password">This is an optional parameter for the basic authentication version. The password used for proxy authentication.</param>
        /// <remarks>Accepted authenticationType values: ["anonymous", "basic"]</remarks>
        /// <returns>
        /// The details of the supplied proxy configuration as a string on success. An exception will be thrown on error.
        /// </returns>
        public static string ConfigureProxy(bool enabled, string address, int port, List<string> noProxyAddresses, string authenticationType, string username = "", string password = "")
        {
            if (string.IsNullOrEmpty(address))
            {
                throw new ValidationError("Cannot configure proxy, address was null or empty!");
            }

            string[] validAuthenticationTypes = new string[] { "anonymous", "basic" };

            if (string.IsNullOrEmpty(authenticationType) || !Array.Exists(validAuthenticationTypes, elem => elem.Equals(authenticationType)))
            {
                throw new ValidationError("Cannot configure proxy, authenticationType is invalid!");
            }

            string response;
            // Check if username and password given for proxy 
            if (!string.IsNullOrEmpty(username) && !string.IsNullOrEmpty(password))
            {
                if (authenticationType.Equals("anonymous"))
                {
                    throw new ValidationError("Cannot configure proxy, incorrect configuration of authentication type and proxy credentials!");
                }
                // Basic type network proxy
                var jsonBody = new { enabled, address, port, noProxyAddresses, authenticationType, username, password };
                response = RestUtils.PostRequest("net/proxy", jsonBody);
            }
            else
            {
                if (authenticationType.Equals("basic") || !string.IsNullOrEmpty(username) || !string.IsNullOrEmpty(password))
                {
                    throw new ValidationError("Cannot configure proxy, incorrect configuration of authentication type and proxy credentials!");
                }
                // Anonymous type network proxy 
                var jsonBody = new { enabled, address, port, noProxyAddresses, authenticationType };
                response = RestUtils.PostRequest("net/proxy", jsonBody);
            }
            return response;
        }

        /// <summary>Makes a "DELETE" request to delete the current network proxy for an attached device.</summary>
        /// <returns>An empty response as a string on success. An exception will be thrown on error.</returns>
        public static string DeleteNetworkProxy()
        {
            return RestUtils.DeleteRequest("net/proxy");
        }

        /// <summary>Makes a "GET" request to retrieve information about the current network proxy for an attached device.</summary>
        /// <returns>The details of the current network proxy as a string on success. An exception will be thrown on error.</returns>
        public static string GetNetworkProxy()
        {
            return RestUtils.GetRequest("net/proxy");
        }

        /// <summary>Makes a "GET" request to retrieve all the failed network connection attempts by an attached device.</summary>
        /// <returns>The failed connection attempts as a string on success. An exception will be thrown on error.</returns>
        public static string GetAllFailedNetworkConnections()
        {
            return RestUtils.GetRequest("wifi/diagnostics/networks");
        }

        /// <summary>Makes a "GET" request to retrieve the status of all network interfaces on an attached device.</summary>
        /// <returns>The status of all network interfaces as a string on success. An exception will be thrown on error.</returns>
        public static string GetAllNetworkInterfaces()
        {
            SinceDeviceAPIVersion.ValidateDeviceApiVersion("GetAllNetworkInterfaces");
            return RestUtils.GetRequest("net/interfaces");
        }

        /// <summary>Makes a "GET" request to retrieve a specific failed network connection attempt by an attached device.</summary>
        /// <param name="networkId">The id of the network you would like the failed connection attempt of.</param>
        /// <returns>The failed network connection attempt as a string on success. An exception will be thrown on error.</returns>
        public static string GetFailedNetworkConnection(string networkId)
        {
            return RestUtils.GetRequest($"wifi/diagnostics/networks/{networkId}");
        }

        /// <summary>Makes a "GET" request to retrieve all network firewall rulesets on an attached device.</summary>
        /// <returns>All network firewall rulesets as a string on success. An exception will be thrown on error.</returns>
        public static string GetNetworkFirewallRuleset()
        {
            SinceDeviceAPIVersion.ValidateDeviceApiVersion("GetNetworkFirewallRuleset");
            return RestUtils.GetRequest("net/firewall/rulesets");
        }

        /// <summary>Makes a "GET" request to retrieve the status of the named network interface on an attached device.</summary>
        /// <param name="networkInterfaceName">The name of the network interface you would like to get the status of.</param>
        /// <returns>The status of the named network interface as a string on success. An exception will be thrown on error.</returns>
        public static string GetNetworkInterface(string networkInterfaceName)
        {
            SinceDeviceAPIVersion.ValidateDeviceApiVersion("GetNetworkInterface");
            if (string.IsNullOrEmpty(networkInterfaceName))
            {
                throw new ValidationError("Cannot get network interface, input was null or empty!");
            }

            return RestUtils.GetRequest($"net/interfaces/{networkInterfaceName}");
        }

        /// <summary>Makes a "GET" request to retrieve the network status on an attached device.</summary>
        /// <returns>The status of the current network as a string on success. An exception will be thrown on error.</returns>
        public static string GetNetworkStatus()
        {
            SinceDeviceAPIVersion.ValidateDeviceApiVersion("GetNetworkStatus");
            return RestUtils.GetRequest("net/status");
        }

        /// <summary>Makes a "PATCH" request to add/update a specific network interface's attributes on an attached device.</summary>
        /// <param name="networkInterfaceName">The name of the network interface you would like to get the status of.</param>
        /// <param name="interfaceUp">If you would like the interface to be active or not.</param>
        /// <returns>An empty response as a string on success. An exception will be thrown on error.</returns>
        public static string SetNetworkInterfaces(string networkInterfaceName, bool interfaceUp)
        {
            SinceDeviceAPIVersion.ValidateDeviceApiVersion("SetNetworkInterfaces");
            if (string.IsNullOrEmpty(networkInterfaceName))
            {
                throw new ValidationError("Cannot set network interface, networkInterfaceName was null or empty!");
            }

            return RestUtils.PatchRequest($"net/interfaces/{networkInterfaceName}", new { interfaceUp });
        }
    }
    /// <summary>Helper class, providing options for ConfigureProxy</summary>
    public static class ConfigureProxyAuthenticationTypeOptions
    {
        /// <summary>anonymous option</summary>
        public static string anonymous = "anonymous";
        /// <summary>basic option</summary>
        public static string basic = "basic";
    }
}
